apps/memerr/Main.hs

```diff
@@ -0,0 +1,48 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StaticPointers #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Main where
+
+import Control.Distributed.Closure
+import Control.Distributed.Spark as RDD
+import Control.Monad
+import Data.Choice 
+import Data.IORef
+import Data.Text (Text)
+import Options.Applicative as Opt
+
+-- Parser for the command line options
+--
+-- the -n option controls how many references to create 
+-- NOTE: 200,000 is about the number we need to crash the JVM with the 
+-- driver memory set to 512M
+--
+-- and if the --no-io-ref flag is present, we will not store references in an
+-- ioref, so the GHC garbage collecter will have a chance to collect them
+argsParser :: Parser (Int, Choice "useIORef")
+argsParser = (,) <$> option auto (value 100 <> Opt.short 'n' <> metavar "N")
+                 <*> flag (Do #useIORef) (Don't #useIORef) (Opt.long "no-io-ref")
+
+main :: IO ()
+main = forwardUnhandledExceptionsToSpark $ do
+    conf <- newSparkConf "Memory memery"
+    -- NOTE: it would be ideal if we could just set the driver memory here
+    -- by dynamically modifying the spark config, but when run in local mode,
+    -- you have to set it through the `spark-submit` CLI
+    sc   <- getOrCreateSparkContext conf
+    (numRefs, useIORef) <- execParser (info argsParser fullDesc)
+    putStrLn $ "# of references to be created: " ++ show numRefs
+    rdd  <- parallelize sc ["yes", "no", "maybe"]
```

Comment:
This RDD is very small. It would be interesting to have a larger one with a few thousand rows.

Reply:
See above comment for motivation for small RDD.

But also, I think it could be interesting to write another program that performs a constant (or relatively smaller) number of operations on a large RDD (which involve several `collects` or reduces) in a linear sequence to demonstrate a situation in which linear types allow us to only keep a single copy of the RDD in the driver memory, but the normal program needs to retain multiple copies (potentially crashing the driver).

Reply:
Using such a small dataset gives more work to the Haskell garbage collector to manage more references, while I think you are interested in having only the JVM garbage collector busy. But we can keep this program for comparison.